In [1]:
# Necessary imports for this notebook
import os

import numpy as np
import pandas as pd

import datetime
import time

import random

# For plotting
%matplotlib inline

import matplotlib.pyplot as plt
import seaborn as sns

sns.set_style('darkgrid', {'axes.facecolor': '0.9'})

## 1. Customer Profiles Generation

Each customer will be defined by the following properties:

* `CUSTOMER_ID`: The customer unique ID
* (`x_customer_id`,`y_customer_id`): A pair of real coordinates (`x_customer_id`,`y_customer_id`) in a 100 * 100 grid, that defines the geographical location of the customer
* (`mean_amount`, `std_amount`):  The mean and standard deviation of the transaction amounts for the customer, assuming that the transaction amounts follow a normal distribution. The `mean_amount` will be drawn from a uniform distribution (5,100) and the `std_amount` will be set as the `mean_amount` divided by two. 
* `mean_nb_tx_per_day`:  The average number of transactions per day for the customer, assuming that the number of transactions per day follows a Poisson distribution. This number will be drawn from a uniform distribution (0,4). 

The `generate_customer_profiles_table` function provides an implementation for generating a table of customer profiles. It takes as input the number of customers for which to generate a profile and a random state for reproducibility. It returns a DataFrame containing the properties for each customer. 

In [2]:
def generate_customer_profiles_table(n_customers, random_state=0):
    
    np.random.seed(random_state)
        
    customer_id_properties=[]
    
    # Generate customer properties from random distributions 
    for customer_id in range(n_customers):
        
        x_customer_id = np.random.uniform(0,100)
        y_customer_id = np.random.uniform(0,100)
        
        mean_amount = np.random.uniform(5,100) # Arbitrary (but sensible) value 
        std_amount = mean_amount/2 # Arbitrary (but sensible) value
        
        mean_nb_tx_per_day = np.random.uniform(0,4) # Arbitrary (but sensible) value 
        
        customer_id_properties.append([customer_id,
                                      x_customer_id, y_customer_id,
                                      mean_amount, std_amount,
                                      mean_nb_tx_per_day])
        
    customer_profiles_table = pd.DataFrame(customer_id_properties, columns=['CUSTOMER_ID',
                                                                      'x_customer_id', 'y_customer_id',
                                                                      'mean_amount', 'std_amount',
                                                                      'mean_nb_tx_per_day'])
    
    return customer_profiles_table

In [3]:
n_customers = 100
customer_profiles_table = generate_customer_profiles_table(n_customers, random_state = 0)
customer_profiles_table

CUSTOMER_ID  x_customer_id  y_customer_id  mean_amount  std_amount  \
0             0      54.881350      71.518937    62.262521   31.131260   
1             1      42.365480      64.589411    46.570785   23.285393   
2             2      96.366276      38.344152    80.213879   40.106939   
3             3      56.804456      92.559664    11.748426    5.874213   
4             4       2.021840      83.261985    78.924891   39.462446   
..          ...            ...            ...          ...         ...   
95           95      85.772264      45.722345    95.428075   47.714038   
96           96      82.076712      90.884372    82.474763   41.237381   
97           97      62.889844      39.843426    10.957730    5.478865   
98           98      25.868407      84.903831     8.163940    4.081970   
99           99      35.536885      35.670689     6.551208    3.275604   

    mean_nb_tx_per_day  
0             2.179533  
1             3.567092  
2             2.115580  
3             0.348517  
4             3.480049  
..                 ...  
95            2.303005  
96            0.637658  
97            1.696129  
98            3.835931  
99            0.740929  

[100 rows x 6 columns]

## 2. Terminal profiles generation

Each terminal will be defined by the following properties:

* `TERMINAL_ID`: The terminal ID
* (`x_terminal_id`,`y_terminal_id`): A pair of real coordinates (`x_terminal_id`,`y_terminal_id`) in a 100 * 100 grid, that defines the geographical location of the terminal

The `generate_terminal_profiles_table` function provides an implementation for generating a table of terminal profiles. It takes as input the number of terminals for which to generate a profile and a random state for reproducibility. It returns a DataFrame containing the properties for each terminal. 


In [4]:
def generate_terminal_profiles_table(n_terminals, random_state=0):
    
    np.random.seed(random_state)
        
    terminal_id_properties=[]
    
    # Generate terminal properties from random distributions 
    for terminal_id in range(n_terminals):
        
        x_terminal_id = np.random.uniform(0,100)
        y_terminal_id = np.random.uniform(0,100)
        
        terminal_id_properties.append([terminal_id,
                                      x_terminal_id, y_terminal_id])
                                       
    terminal_profiles_table = pd.DataFrame(terminal_id_properties, columns=['TERMINAL_ID',
                                                                      'x_terminal_id', 'y_terminal_id'])
    
    return terminal_profiles_table

In [5]:
n_terminals = 10
terminal_profiles_table = generate_terminal_profiles_table(n_terminals, random_state = 0)
terminal_profiles_table

TERMINAL_ID  x_terminal_id  y_terminal_id
0            0      54.881350      71.518937
1            1      60.276338      54.488318
2            2      42.365480      64.589411
3            3      43.758721      89.177300
4            4      96.366276      38.344152
5            5      79.172504      52.889492
6            6      56.804456      92.559664
7            7       7.103606       8.712930
8            8       2.021840      83.261985
9            9      77.815675      87.001215

## 3. Association of customer profiles to terminals

At the stage, the terminals will be associated with the appropraite customer profiles. In this design, customers can only perform transactions on terminals that are within a radius of `r` of their geographical locations. 

A function, called `get_list_terminals_within_radius`, was defined to find these terminals for a customer profile. The function will take as input a customer profile (any row in the customer profiles table), an array that contains the geographical location of all terminals, and the radius `r`. It will return the list of terminals within a radius of `r` for that customer. 

In [6]:
def get_list_terminals_within_radius(customer_profile, x_y_terminals, r):
    
    # Use numpy arrays in the following to speed up computations
    
    # Location (x,y) of customer as numpy array
    x_y_customer = customer_profile[['x_customer_id','y_customer_id']].values.astype(float)
    
    # Squared difference in coordinates between customer and terminal locations
    squared_diff_x_y = np.square(x_y_customer - x_y_terminals)
    
    # Sum along rows and compute suared root to get distance
    dist_x_y = np.sqrt(np.sum(squared_diff_x_y, axis=1))
    
    # Get the indices of terminals which are at a distance less than r
    available_terminals = list(np.where(dist_x_y<r)[0])
    
    # Return the list of terminal IDs
    return available_terminals
    

In [7]:
# We first get the geographical locations of all terminals as a numpy array
x_y_terminals = terminal_profiles_table[['x_terminal_id','y_terminal_id']].values.astype(float)

In [8]:
customer_profiles_table['available_terminals']=customer_profiles_table.apply(lambda x : get_list_terminals_within_radius(x, x_y_terminals=x_y_terminals, r=50), axis=1)
customer_profiles_table

CUSTOMER_ID  x_customer_id  y_customer_id  mean_amount  std_amount  \
0             0      54.881350      71.518937    62.262521   31.131260   
1             1      42.365480      64.589411    46.570785   23.285393   
2             2      96.366276      38.344152    80.213879   40.106939   
3             3      56.804456      92.559664    11.748426    5.874213   
4             4       2.021840      83.261985    78.924891   39.462446   
..          ...            ...            ...          ...         ...   
95           95      85.772264      45.722345    95.428075   47.714038   
96           96      82.076712      90.884372    82.474763   41.237381   
97           97      62.889844      39.843426    10.957730    5.478865   
98           98      25.868407      84.903831     8.163940    4.081970   
99           99      35.536885      35.670689     6.551208    3.275604   

    mean_nb_tx_per_day       available_terminals  
0             2.179533     [0, 1, 2, 3, 5, 6, 9]  
1             3.567092  [0, 1, 2, 3, 5, 6, 8, 9]  
2             2.115580                 [1, 4, 5]  
3             0.348517     [0, 1, 2, 3, 5, 6, 9]  
4             3.480049                 [2, 3, 8]  
..                 ...                       ...  
95            2.303005        [0, 1, 2, 4, 5, 9]  
96            0.637658     [0, 1, 2, 3, 5, 6, 9]  
97            1.696129        [0, 1, 2, 4, 5, 9]  
98            3.835931        [0, 1, 2, 3, 6, 8]  
99            0.740929           [0, 1, 2, 5, 7]  

[100 rows x 7 columns]

## 4. Generation of transactions

The customer profiles now contain all the information that we require to generate transactions. The transaction generation will be done by a function `generate_transactions_table` that takes as input a customer profile, a starting date, and a number of days for which to generate transactions. It will return a table of transactions. 

In [9]:
def generate_transactions_table(customer_profile, start_date = "2022-01-01", nb_days = 10):
    
    customer_transactions = []
    
    random.seed(int(customer_profile.CUSTOMER_ID))
    np.random.seed(int(customer_profile.CUSTOMER_ID))
    
    # For all days
    for day in range(nb_days):
        
        # Random number of transactions for that day 
        nb_tx = np.random.poisson(customer_profile.mean_nb_tx_per_day)
        
        # If nb_tx positive, let us generate transactions
        if nb_tx>0:
            
            for tx in range(nb_tx):
                
                # Time of transaction: Around noon, std 20000 seconds. This choice aims at simulating the fact that 
                # most transactions occur during the day.
                time_tx = int(np.random.normal(86400/2, 20000))
                
                # If transaction time between 0 and 86400, let us keep it, otherwise, let us discard it
                if (time_tx>0) and (time_tx<86400):
                    
                    # Amount is drawn from a normal distribution  
                    amount = np.random.normal(customer_profile.mean_amount, customer_profile.std_amount)
                    
                    # If amount negative, draw from a uniform distribution
                    if amount<0:
                        amount = np.random.uniform(0,customer_profile.mean_amount*2)
                    
                    amount=np.round(amount,decimals=2)
                    
                    if len(customer_profile.available_terminals)>0:
                        
                        terminal_id = random.choice(customer_profile.available_terminals)
                    
                        customer_transactions.append([time_tx+day*86400, day,
                                                      customer_profile.CUSTOMER_ID, 
                                                      terminal_id, amount])
            
    customer_transactions = pd.DataFrame(customer_transactions, columns=['TX_TIME_SECONDS', 'TX_TIME_DAYS', 'CUSTOMER_ID', 'TERMINAL_ID', 'TX_AMOUNT'])
    
    if len(customer_transactions)>0:
        customer_transactions['TX_DATETIME'] = pd.to_datetime(customer_transactions["TX_TIME_SECONDS"], unit='s', origin=start_date)
        customer_transactions=customer_transactions[['TX_DATETIME','CUSTOMER_ID', 'TERMINAL_ID', 'TX_AMOUNT','TX_TIME_SECONDS', 'TX_TIME_DAYS']]
    
    return customer_transactions  

In [10]:
transactions_df=customer_profiles_table.groupby('CUSTOMER_ID').apply(lambda x : generate_transactions_table(x.iloc[0], nb_days=5)).reset_index(drop=True)
transactions_df

TX_DATETIME CUSTOMER_ID TERMINAL_ID  TX_AMOUNT TX_TIME_SECONDS  \
0   2022-01-01 07:19:05           0           9     123.59           26345   
1   2022-01-01 19:02:02           0           3      46.51           68522   
2   2022-01-01 18:00:16           0           9      77.34           64816   
3   2022-01-02 15:13:02           0           3      32.35          141182   
4   2022-01-02 14:05:38           0           0      63.30          137138   
..                  ...         ...         ...        ...             ...   
872 2022-01-05 07:15:38          98           1       5.68          371738   
873 2022-01-05 10:34:56          98           3       5.15          383696   
874 2022-01-05 08:49:22          98           0       3.39          377362   
875 2022-01-01 13:34:25          99           5      10.91           48865   
876 2022-01-04 16:11:43          99           5       9.26          317503   

    TX_TIME_DAYS  
0              0  
1              0  
2              0  
3              1  
4              1  
..           ...  
872            4  
873            4  
874            4  
875            0  
876            3  

[877 rows x 6 columns]

<h3> Scaling up to a larger dataset</h3>

All the building blocks to generate a larger dataset are avaliable. A function `generate_dataset` will be defined, that will take care of running all the previous steps. It will 

* take as inputs the number of desired customers, terminals and days, as well as the starting date and the radius `r`
* return the generated customer and terminal profiles table, and the DataFrame of transactions.

In [11]:
def generate_dataset(n_customers = 10000, n_terminals = 1000000, nb_days=90, start_date="2022-01-01", r=5):
    
    start_time=time.time()
    customer_profiles_table = generate_customer_profiles_table(n_customers, random_state = 0)
    print("Time to generate customer profiles table: {0:.2}s".format(time.time()-start_time))
    
    start_time=time.time()
    terminal_profiles_table = generate_terminal_profiles_table(n_terminals, random_state = 1)
    print("Time to generate terminal profiles table: {0:.2}s".format(time.time()-start_time))
    
    start_time=time.time()
    x_y_terminals = terminal_profiles_table[['x_terminal_id','y_terminal_id']].values.astype(float)
    customer_profiles_table['available_terminals'] = customer_profiles_table.apply(lambda x : get_list_terminals_within_radius(x, x_y_terminals=x_y_terminals, r=r), axis=1)
    # With Pandarallel
    #customer_profiles_table['available_terminals'] = customer_profiles_table.parallel_apply(lambda x : get_list_closest_terminals(x, x_y_terminals=x_y_terminals, r=r), axis=1)
    customer_profiles_table['nb_terminals']=customer_profiles_table.available_terminals.apply(len)
    print("Time to associate terminals to customers: {0:.2}s".format(time.time()-start_time))
    
    start_time=time.time()
    transactions_df=customer_profiles_table.groupby('CUSTOMER_ID').apply(lambda x : generate_transactions_table(x.iloc[0], nb_days=nb_days)).reset_index(drop=True)
    # With Pandarallel
    #transactions_df=customer_profiles_table.groupby('CUSTOMER_ID').parallel_apply(lambda x : generate_transactions_table(x.iloc[0], nb_days=nb_days)).reset_index(drop=True)
    print("Time to generate transactions: {0:.2}s".format(time.time()-start_time))
    
    # Sort transactions chronologically
    transactions_df=transactions_df.sort_values('TX_DATETIME')
    # Reset indices, starting from 0
    transactions_df.reset_index(inplace=True,drop=True)
    transactions_df.reset_index(inplace=True)
    # TRANSACTION_ID are the dataframe indices, starting from 0
    transactions_df.rename(columns = {'index':'TRANSACTION_ID'}, inplace = True)
    
    return (customer_profiles_table, terminal_profiles_table, transactions_df)
    

A dataset with the following features, will be generated: 

* 10000 customers
* 20000 terminals
* 150 days of transactions (which corresponds to a simulated period from 2022/01/01 to 2022/05/31)

The starting date is arbitrarily fixed at 2022/01/01. The radius $r$ is set to 5, which corresponds to around 100 available terminals for each customer.

In [12]:
(customer_profiles_table, terminal_profiles_table, transactions_df)=\
    generate_dataset(n_customers = 10000, 
                     n_terminals = 20000, 
                     nb_days=151, 
                     start_date="2022-01-01", 
                     r=5)


Time to generate customer profiles table: 0.1s
Time to generate terminal profiles table: 0.094s
Time to associate terminals to customers: 5.0s
Time to generate transactions: 2.3e+02s


In [13]:
transactions_df.shape

(2928154, 7)

In [14]:
transactions_df

TRANSACTION_ID         TX_DATETIME CUSTOMER_ID TERMINAL_ID  \
0                     0 2022-01-01 00:00:17        6160        6202   
1                     1 2022-01-01 00:00:31         596        6532   
2                     2 2022-01-01 00:01:05        9339       10651   
3                     3 2022-01-01 00:02:10        4961        8673   
4                     4 2022-01-01 00:02:21        6170        4884   
...                 ...                 ...         ...         ...   
2928149         2928149 2022-05-31 23:59:38        8127        3133   
2928150         2928150 2022-05-31 23:59:43        9916        1138   
2928151         2928151 2022-05-31 23:59:46        2538       11932   
2928152         2928152 2022-05-31 23:59:50        8113       18820   
2928153         2928153 2022-05-31 23:59:55        8539       17831   

         TX_AMOUNT TX_TIME_SECONDS TX_TIME_DAYS  
0            31.83              17            0  
1            57.16              31            0  
2            28.92              65            0  
3            81.51             130            0  
4            25.17             141            0  
...            ...             ...          ...  
2928149      39.82        13046378          150  
2928150     103.88        13046383          150  
2928151      47.92        13046386          150  
2928152     117.70        13046390          150  
2928153     203.29        13046395          150  

[2928154 rows x 7 columns]

## 5. Fraud scenarios generation

This last step of the simulation adds fraudulent transactions to the dataset, using the following fraud scenarios:

* Scenario 1: Any transaction whose amount is more than 220 is a fraud. This scenario is not inspired by a real-world scenario. Rather, it will provide an obvious fraud pattern that should be detected by any baseline fraud detector. This will be useful to validate the implementation of a fraud detection technique.  

* Scenario 2: Every day, a list of two terminals is drawn at random. All transactions on these terminals in the next 28 days will be marked as fraudulent. This scenario simulates a criminal use of a terminal, through phishing for example. Detecting this scenario will be possible by adding features that keep track of the number of fraudulent transactions on the terminal. Since the terminal is only compromised for 28 days, additional strategies that involve concept drift will need to be designed to efficiently deal with this scenario.     

* Scenario 3: Every day, a list of 3 customers is drawn at random. In the next 14 days, 1/3 of their transactions have their amounts multiplied by 5 and marked as fraudulent. This scenario simulates a card-not-present fraud where the credentials of a customer have been leaked. The customer continues to make transactions, and transactions of higher values are made by the fraudster who tries to maximize their gains. Detecting this scenario will require adding features that keep track of the spending habits of the customer. As for scenario 2, since the card is only temporarily compromised, additional strategies that involve concept drift should also be designed. 



In [15]:
def add_frauds(customer_profiles_table, terminal_profiles_table, transactions_df):
    
    # By default, all transactions are genuine
    transactions_df['TX_FRAUD']=0
    transactions_df['TX_FRAUD_SCENARIO']=0
    
    # Scenario 1
    transactions_df.loc[transactions_df.TX_AMOUNT>220, 'TX_FRAUD']=1
    transactions_df.loc[transactions_df.TX_AMOUNT>220, 'TX_FRAUD_SCENARIO']=1
    nb_frauds_scenario_1=transactions_df.TX_FRAUD.sum()
    print("Number of frauds from scenario 1: "+str(nb_frauds_scenario_1))
    
    # Scenario 2
    for day in range(transactions_df.TX_TIME_DAYS.max()):
        
        compromised_terminals = terminal_profiles_table.TERMINAL_ID.sample(n=2, random_state=day)
        
        compromised_transactions=transactions_df[(transactions_df.TX_TIME_DAYS>=day) & 
                                                    (transactions_df.TX_TIME_DAYS<day+28) & 
                                                    (transactions_df.TERMINAL_ID.isin(compromised_terminals))]
                            
        transactions_df.loc[compromised_transactions.index,'TX_FRAUD']=1
        transactions_df.loc[compromised_transactions.index,'TX_FRAUD_SCENARIO']=2
    
    nb_frauds_scenario_2=transactions_df.TX_FRAUD.sum()-nb_frauds_scenario_1
    print("Number of frauds from scenario 2: "+str(nb_frauds_scenario_2))
    
    # Scenario 3
    for day in range(transactions_df.TX_TIME_DAYS.max()):
        
        compromised_customers = customer_profiles_table.CUSTOMER_ID.sample(n=3, random_state=day).values
        
        compromised_transactions=transactions_df[(transactions_df.TX_TIME_DAYS>=day) & 
                                                    (transactions_df.TX_TIME_DAYS<day+14) & 
                                                    (transactions_df.CUSTOMER_ID.isin(compromised_customers))]
        
        nb_compromised_transactions=len(compromised_transactions)
        
        
        random.seed(day)
        index_fauds = random.sample(list(compromised_transactions.index.values),k=int(nb_compromised_transactions/3))
        
        transactions_df.loc[index_fauds,'TX_AMOUNT']=transactions_df.loc[index_fauds,'TX_AMOUNT']*5
        transactions_df.loc[index_fauds,'TX_FRAUD']=1
        transactions_df.loc[index_fauds,'TX_FRAUD_SCENARIO']=3
        
                             
    nb_frauds_scenario_3=transactions_df.TX_FRAUD.sum()-nb_frauds_scenario_2-nb_frauds_scenario_1
    print("Number of frauds from scenario 3: "+str(nb_frauds_scenario_3))
    
    return transactions_df                 


In [16]:
%time transactions_df = add_frauds(customer_profiles_table, terminal_profiles_table, transactions_df)

Number of frauds from scenario 1: 1636
Number of frauds from scenario 2: 7378
Number of frauds from scenario 3: 4089
Wall time: 2min 40s


Percentage of fraudulent transactions:

In [17]:
transactions_df.TX_FRAUD.mean()*100

0.44748329493599037

Number of fraudulent transactions:

In [18]:
transactions_df.TX_FRAUD.sum()

13103

A total of `13103`  transactions were marked as fraudulent. This amounts to 0.45% of the transactions. Note that the sum of the frauds for each scenario does not equal the total amount of fraudulent transactions. This is because the same transactions may have been marked as fraudulent by two or more fraud scenarios.  

The simulated transaction dataset is now complete, with a fraudulent label added to all transactions.

In [19]:
transactions_df.head()

TRANSACTION_ID         TX_DATETIME CUSTOMER_ID TERMINAL_ID  TX_AMOUNT  \
0               0 2022-01-01 00:00:17        6160        6202      31.83   
1               1 2022-01-01 00:00:31         596        6532      57.16   
2               2 2022-01-01 00:01:05        9339       10651      28.92   
3               3 2022-01-01 00:02:10        4961        8673      81.51   
4               4 2022-01-01 00:02:21        6170        4884      25.17   

  TX_TIME_SECONDS TX_TIME_DAYS  TX_FRAUD  TX_FRAUD_SCENARIO  
0              17            0         0                  0  
1              31            0         0                  0  
2              65            0         0                  0  
3             130            0         0                  0  
4             141            0         0                  0

In [24]:
transactions_df.drop(['TX_FRAUD_SCENARIO'], axis = 1, inplace = True)
transactions_df.head()

TRANSACTION_ID         TX_DATETIME CUSTOMER_ID TERMINAL_ID  TX_AMOUNT  \
0               0 2022-01-01 00:00:17        6160        6202      31.83   
1               1 2022-01-01 00:00:31         596        6532      57.16   
2               2 2022-01-01 00:01:05        9339       10651      28.92   
3               3 2022-01-01 00:02:10        4961        8673      81.51   
4               4 2022-01-01 00:02:21        6170        4884      25.17   

  TX_TIME_SECONDS TX_TIME_DAYS  TX_FRAUD  
0              17            0         0  
1              31            0         0  
2              65            0         0  
3             130            0         0  
4             141            0         0

In [28]:
transactions_df.tail()

TRANSACTION_ID         TX_DATETIME CUSTOMER_ID TERMINAL_ID  \
2928149         2928149 2022-05-31 23:59:38        8127        3133   
2928150         2928150 2022-05-31 23:59:43        9916        1138   
2928151         2928151 2022-05-31 23:59:46        2538       11932   
2928152         2928152 2022-05-31 23:59:50        8113       18820   
2928153         2928153 2022-05-31 23:59:55        8539       17831   

         TX_AMOUNT TX_TIME_SECONDS TX_TIME_DAYS  TX_FRAUD  
2928149      39.82        13046378          150         0  
2928150     103.88        13046383          150         0  
2928151      47.92        13046386          150         0  
2928152     117.70        13046390          150         0  
2928153     203.29        13046395          150         0

In [25]:
transactions_df.to_pickle('simulated-data-raw')

In [26]:
transactions_df.to_csv('simulated-data-raw.csv')